### PR TITLE
[DO NOT MERGE] scenario-based interface tester

### PR DESCRIPTION
**Pull request by PietroPasotti**
This is a WIP/POC thing, currently being broken down into smaller PRs. DO NOT MERGE this branch, but also do not delete it just yet.

To see what the interface tests would look like from the perspective of a charm implementing one of the interfaces, using the pytest plugin proposed in this PR:
see https://github.com/canonical/traefik-k8s-operator/tree/interface-tests

### CODE REVIEW

File: tester/plugin/pytest_interface_tester.py [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+                    f"interface spec dir not found at expected location. "
+                    f"Check that {repo_name}/{self._base_path}/{interface_name} "
+                    f"is a valid path in the remote repo you've selected as test case source "
+                    f"for the plugin."
+                )
+
+            tests = gather_test_spec_for_version(intf_spec_path)
+        return tests
+
+    def _yield_tests(
```
[the remark this answers is not in the document]
Comment (gruyaume) [reply tester/plugin/pytest_interface_tester.py:151]: Also, as mentioned, it's pretty complex, I think we should try to optimise for readability here.

File: tester/plugin/pytest_interface_tester.py [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+        try:
+            test.validate_schema(relation, schema=schema)
+        except Exception as e:
+            msg = f"Failed check 3: validating schema on scene output: {e}"
+            logger.error(msg)
+            raise RuntimeError(msg) from e
+
+    def _assert_schemas_valid(
+        self, test: "InterfaceTestCase", schema: "DataBagSchema", state_out: State
+    ):
```
Comment (gruyaume) [inline tester/plugin/pytest_interface_tester.py:350]: Add return typing

File: tester/plugin/pytest_interface_tester.py [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+                errors.append(e.args[0])
+        return errors
+
+    def _run(
+        self,
+        test: "InterfaceTestCase",
+        schema: Optional["DataBagSchema"],
+        scene: Scene,
+        subtests=None,
+    ):
```
Comment (gruyaume) [inline tester/plugin/pytest_interface_tester.py:368]: Add return typing

**Review by gruyaume — COMMENTED** [review]
I'm aware this is still in draft but wanted to look at what's being done here. I like the direction a lot :)  In general I'd recommend to keep things simple and add unit tests. We'll have to test our interface tester if we don't want future changes to break it. This is by no mean a complete code review, more like initial nitpicks and minor comments.

File: tester/plugin/pytest_interface_tester.py [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+        base_path: str = "interfaces",
+    ):
+        self._repo = repo
+        self._branch = branch
+        self._base_path = base_path
+        self._target = None
+        self._interface_name = None
+        self._state_template = None
+
+    def configure(
```
Comment (gruyaume) [inline tester/plugin/pytest_interface_tester.py:56]: Why this `configure` in general? Can't we leverage `__init__`?

File: tester/plugin/pytest_interface_tester.py [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+                logger.info(f"collected {scene} for {interface_name}")
+
+                schema = spec["schema"]
+                yield test, schema, scene
+
+    def run(self, subtests=None):
+        """Run interface tests."""
+        self._check_config()  # will raise if misconfigured
+
+        interface_name = self._interface_name
```
Comment (gruyaume) [inline tester/plugin/pytest_interface_tester.py:304]: Setting this here doesn't seem necessary

File: tester/plugin/pytest_interface_tester.py [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+    def _assert_state_out_valid(self, state: State, test: "InterfaceTestCase"):
+        try:
+            # todo consider validate(output: RelationSpec) and INPUT_STATE: RelationSpec to simplify.
+            test.validate(state)
+        except Exception as e:
+            msg = f"Failed check 2: validating scene output: {e}"
+            raise RuntimeError(msg) from e
+
+    def _assert_schema_valid(
+        self, test: "InterfaceTestCase", relation, schema: "DataBagSchema"
```
Comment (gruyaume) [inline tester/plugin/pytest_interface_tester.py:339]: Add typing for `relation`

File: tester/README.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+If you need to patch your charm for the tests to run, this is where you do it:
+
+```python
+from unittest.mock import patch
+from interface_tester import InterfaceTester
+from charm import MyCharm
+
+def test_all_interfaces(interface_tester: InterfaceTester):
+    with patch('charm.KubernetesAPI', lambda **unused: None):
+      interface_tester.configure(charm=MyCharm)
```
Comment (gruyaume) [inline tester/README.md:33]: `configure` doesn't have a `charm` parameter

File: tester/plugin/pytest_interface_tester.py [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+        self._state_template = None
+
+    def configure(
+        self,
+        *,
+        target: Union[Type[CharmType], CharmSpec] = None,
+        repo: Optional[str] = None,
+        branch: Optional[str] = None,
+        base_path: Optional[str] = None,
+        interface_name: str = None,
```
[the remark this answers is not in the document]
Comment (PietroPasotti) [reply tester/plugin/pytest_interface_tester.py:63]: this is a builder: you can configure() multiple times, each time setting different subsets of the 'total' config.

File: tester/plugin/pytest_interface_tester.py (hunk 4 above)
Comment (PietroPasotti) [reply tester/plugin/pytest_interface_tester.py:56]: the idea is to make `interface_tester` a pytest fixture, which means we need to initialize the object at a time where we don't know what interface we will be testing.

**Pull request closed without merging**